crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
+        let lock_version = Header::read(&mut lock_file)?.version;
+        match toolchain_version {
+            // No ToolchainVersion and old Move.lock version implies legacy compiler.
+            None if lock_version == LEGACY_MOVE_LOCK_VERSION => {
+                debug!("{package} on legacy compiler",);
+                let legacy_toolchain = ToolchainVersion {
+                    compiler_version: LEGACY_COMPILER_VERSION.into(),
+                    edition: Edition::LEGACY,
+                    flavor: Flavor::Sui,
+                };
+                package_version_map.insert(*package, legacy_toolchain);
+            }
+            // No ToolchainVersion and new Move.lock version implies current compiler.
+            None => {
+                debug!("{package} on current compiler @ {CURRENT_COMPILER_VERSION}",);
+                package_version_map.insert(*package, current_toolchain);
+            }
+            // This dependency uses the current compiler.
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == CURRENT_COMPILER_VERSION => {
+                debug!("{package} on current compiler @ {CURRENT_COMPILER_VERSION}",);
+                package_version_map.insert(*package, current_toolchain);
+            }
+            // This dependency needs a prior compiler. Mark it and compile.
+            Some(toolchain_version) => {
+                println!(
+                    "{} {package} compiler @ {}",
+                    "REQUIRE".bold().green(),
+                    toolchain_version.compiler_version.yellow(),
+                );
+                package_version_map.insert(*package, toolchain_version);
+            }
+        }
+    }
+
+    let mut units = vec![];
+    let mut package_compiled: HashSet<Symbol> = HashSet::new();
+    // Iterate over compiled units, and determine whether they need to be recompiled and replaced by a prior compiler.
+    for (package, local_unit) in compiled_units {
+        match package_version_map.get(package) {
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == CURRENT_COMPILER_VERSION => {
+                units.push((*package, local_unit.clone()));
+                continue;
+            }
+            Some(toolchain_version) if !package_compiled.contains(package) => {
+                let package_root = SourcePackageLayout::try_find_root(&local_unit.source_path)?;
+                download_and_compile(package_root.clone(), toolchain_version, package)?;
+                package_compiled.insert(*package);
+
+                let compiled_unit_paths = vec![package_root.clone()];
+                let compiled_units = find_filenames(&compiled_unit_paths, |path| {
+                    extension_equals(path, MOVE_COMPILED_EXTENSION)
+                })?;
+                let build_path = package_root.clone().join("build").join(package.as_str());
```

Comment:
Does this imply that compiling that publishing the root package might modify the build directories of dependency packages because of dependency verification? I'd like to avoid that if we can.  Can we store these separately in the root package's build directory? cc @tzakian who can advise on the abstractions we use to refer to things in the build directory as well -- IIRC we like to avoid referring to the `build` directory by hardcoded name, because its location is configurable.

Reply:
I think I ended up with a nice solution here to avoid polluting the package's build dir, which is to simply compile to a temporary directory using `--instal-dir` 

(we need to compile dependencies anyway, and we only ever do this for dependencies with previous compilers, so tossing it in a tempdir works well I think)

this also fixes referring to `build` manually, and uses the package's typed definition instead (i.e., `CompiledPackageLayout::path(&CompiledPackageLayout::Root`)

cb5ff8a5820439b2d9ede892be13c6232d3925d6